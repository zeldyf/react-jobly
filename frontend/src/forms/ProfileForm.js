import React, { useState, useContext } from 'react';
import { Card, CardBody, Form, Label, Input, Button } from 'reactstrap';

import UserContext from '../UserContext';
import JoblyApi from '../common/api';
import JobsAppliedTo from '../jobs/JobsAppliedTo';

/** Edit profile form.
 *
 * Displays profile form and handles changes to local form state.
 * Submitting the form calls the API to save, and triggers user info reloading
 * throughout the site.
 *
 * Confirmation of a successful save.
 *
 * Routed as /profile
 * Routes -> ProfileForm
 */

const ProfileForm = ({ deleteUser }) => {
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const INITIAL_STATE = {
    username: currentUser.username,
    password: '',
    firstName: currentUser.firstName,
    lastName: currentUser.lastName,
    email: currentUser.email
  };
  const [formData, setFormData] = useState(INITIAL_STATE);
  const [formErrors, setFormErrors] = useState([]);
  const [message, setMessage] = useState('');
  const [isShown, setIsShown] = useState(false);

  console.debug(
    'ProfileForm',
    'currentUser=',
    currentUser,
    'formData=',
    formData,
    'formErrors=',
    formErrors
  );

  const showApps = () => {
    setIsShown(current => !current);
  };

  /** Update form fields */
  const handleChange = evt => {
    const { name, value } = evt.target;
    setFormData(data => ({
      ...data,
      [name]: value
    }));
    setFormErrors([]);
  };

  /** on form submit:
   * - attempt save to backend & report any errors
   * - if successful
   *   - clear previous error messages and password
   *   - show save-confirmed message
   *   - set current user info throughout the site
   */
  const handleSubmit = async evt => {
    evt.preventDefault();

    let profileData = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
      password: formData.password
    };

    let username = formData.username;
    let updatedUser;

    try {
      updatedUser = await JoblyApi.saveProfile(username, profileData);
      setMessage('Saved successfully!');
    } catch (errors) {
      setFormErrors(errors);
      return;
    }

    setFormData(f => ({ ...f, password: '' }));
    setFormErrors([]);

    // trigger reloading of user information throughout the site
    setCurrentUser(updatedUser);
  };

  return (
    <div className="Profile col-md-5 offset-md-4 col-lg-4 offset-lg-4 mt-3">
      <Card>
        <CardBody>
        <h1>Profile</h1>
          <Form className="ProfileForm-form" onSubmit={handleSubmit}>
            <Label htmlFor="username" className="ProfileForm-Label">
              Username
            </Label>
            <p>@{currentUser.username}</p>
            <Label htmlFor="firstName" className="ProfileForm-Label">
              First Name
            </Label>
            <Input
              className="ProfileForm-Input"
              id="firstName"
              name="firstName"
              type="text"
              placeholder="First Name"
              value={formData.firstName}
              onChange={handleChange}
            ></Input>
            <Label htmlFor="lastName" className="ProfileForm-Label">
              Last Name
            </Label>
            <Input
              className="ProfileForm-Input"
              id="lastName"
              name="lastName"
              type="text"
              placeholder="Last Name"
              value={formData.lastName}
              onChange={handleChange}
            ></Input>
            <Label htmlFor="email" className="ProfileForm-Label">
              Email
            </Label>
            <Input
              className="ProfileForm-Input"
              id="email"
              name="email"
              type="text"
              placeholder="Email"
              value={formData.email}
              onChange={handleChange}
            ></Input>
            <Label htmlFor="password" className="ProfileForm-Label">
              Confirm password to make changes:
            </Label>
            <Input
              className="ProfileForm-Input"
              id="password"
              name="password"
              type="text"
              placeholder="Password"
              value={formData.password}
              onChange={handleChange}
              required
            ></Input>
            <span className="ProfileForm-message">
              {formErrors ? <p>{formErrors}</p> : null}
              {message ? <p>{message}</p> : null}
            </span>
            <Button
              type="submit"
              className="btn btn-lg btn-block"
              color="primary"
              outline
            >
              Save Changes
            </Button>
          </Form>
        </CardBody>
      </Card>
      {/* Toggles list of jobs to show on page */}
      <div className="Expand text-center">
        <Button onClick={showApps} className="btn-sm m-3" color="info">
          {isShown ? 'Hide Applied Jobs' : 'Show Applied Jobs'}
        </Button>
      </div>
      {isShown && <JobsAppliedTo />}
      <div className="Delete text-center">
        <button
          onClick={deleteUser}
          className="btn btn-link my-5"
          style={{ color: 'red' }}
        >
          Delete my profile
        </button>
      </div>
    </div>
  );
};

export default ProfileForm;
