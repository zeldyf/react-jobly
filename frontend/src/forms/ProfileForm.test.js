import React from 'react';
import { render } from '@testing-library/react';
import ProfileForm from './ProfileForm';
import UserContext from '../UserContext';

const currentUser = {
  username: 'testuser',
  firstName: 'Test',
  lastName: 'User',
  email: 'testuser@gmail.com',
};

const deleteUser = jest.fn();

// Mock context for testing
const Wrapper = ({ children }) => (
  <UserContext.Provider value={{ currentUser }}>{children}</UserContext.Provider>
);

// Smoke test
it('renders without crashing', () => {
  render(<ProfileForm deleteUser={deleteUser} />, { wrapper: Wrapper });
});

// Snapshot test
it('matches snapshot', () => {
  const { asFragment } = render(<ProfileForm deleteUser={deleteUser} />, {
    wrapper: Wrapper,
  });
  expect(asFragment()).toMatchSnapshot();
});
