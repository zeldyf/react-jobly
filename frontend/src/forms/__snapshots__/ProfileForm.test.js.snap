// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Profile col-md-5 offset-md-4 col-lg-4 offset-lg-4 mt-3"
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <h1>
          Profile
        </h1>
        <form
          class="ProfileForm-form"
        >
          <label
            class="ProfileForm-Label"
            for="username"
          >
            Username
          </label>
          <p>
            @testuser
          </p>
          <label
            class="ProfileForm-Label"
            for="firstName"
          >
            First Name
          </label>
          <input
            class="ProfileForm-Input form-control"
            id="firstName"
            name="firstName"
            placeholder="First Name"
            type="text"
            value="Test"
          />
          <label
            class="ProfileForm-Label"
            for="lastName"
          >
            Last Name
          </label>
          <input
            class="ProfileForm-Input form-control"
            id="lastName"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value="User"
          />
          <label
            class="ProfileForm-Label"
            for="email"
          >
            Email
          </label>
          <input
            class="ProfileForm-Input form-control"
            id="email"
            name="email"
            placeholder="Email"
            type="text"
            value="testuser@gmail.com"
          />
          <label
            class="ProfileForm-Label"
            for="password"
          >
            Confirm password to make changes:
          </label>
          <input
            class="ProfileForm-Input form-control"
            id="password"
            name="password"
            placeholder="Password"
            required=""
            type="text"
            value=""
          />
          <span
            class="ProfileForm-message"
          >
            <p />
          </span>
          <button
            class="btn btn-lg btn-block btn btn-outline-primary"
            type="submit"
          >
            Save Changes
          </button>
        </form>
      </div>
    </div>
    <div
      class="Expand text-center"
    >
      <button
        class="btn-sm m-3 btn btn-info"
        type="button"
      >
        Show Applied Jobs
      </button>
    </div>
    <div
      class="Delete text-center"
    >
      <button
        class="btn btn-link my-5"
        style="color: red;"
      >
        Delete my profile
      </button>
    </div>
  </div>
</DocumentFragment>
`;
